File: Laptop.py
from Producto_electronico import ProductoElectronico
from con_descuento import ConDescuento


class Laptop(ProductoElectronico, ConDescuento):
    def __init__(self, id_producto, nombre, marca, precio, stock, procesador, ram_gb, es_gamer):
        super().__init__(id_producto, nombre, marca, precio, stock)
        self.procesador = procesador
        self.ram_gb = ram_gb
        self.es_gamer = es_gamer
        self.descuento_aplicado = True

    def mostrar_informacion(self):
        print(f" {self.nombre} ({self.marca}) - {self.procesador} - RAM: {self.ram_gb}GB - Gamer: {'Sí' if self.es_gamer else 'No'} - Precio: €{self.precio:.2f} - Stock: {self.stock}")

    def actualizar_precio(self, porcentaje):
        self.precio *= (1 + porcentaje / 100)

    def aplicar_descuento(self, porcentaje):
        if porcentaje < 0 or porcentaje > 100:
            print("Revisa: Porcentaje de descuento inválido.")
            return
        self.precio *= (1 - porcentaje / 100)
        self.descuento_aplicado = True

    def mostrar_precio_con_descuento(self):
        print(f"{self.nombre} (Laptop) con descuento: €{self.precio:.2f}")

File: Laptop.java
package SistemaGestionTienda;

public class Laptop extends ProductoElectronico implements ConDescuento {
    private String procesador;
    private int RAM_GB;
    private boolean esGamer;

    // Constructor con sobrecarga
    public Laptop(String idProducto, String nombre, String marca,
              double precio, int stock,
              String procesador, int RAM_GB, boolean esGamer) {
    super(idProducto, nombre, marca, precio, stock);
    this.procesador = procesador;
    this.RAM_GB = RAM_GB;
    this.esGamer = esGamer;
}
    
   // Getters y Setters
    public String getProcesador() {
        return procesador;
    }

    public void setProcesador(String procesador) {
        this.procesador = procesador;
    }

    public int getRAM_GB() {
        return RAM_GB;
    }

    public void setRAM_GB(int RAM_GB) {
        this.RAM_GB = RAM_GB;
    }

    public boolean isEsGamer() {
        return esGamer;
    }

    public void setEsGamer(boolean esGamer) {
        this.esGamer = esGamer;
    }

    // Implementación de métodos abstractos
    @Override
    public void mostrarInformacion() {
        System.out.println("Laptop: " + getNombre());
        System.out.println("Marca: " + getMarca());
        System.out.println("Procesador: " + procesador);
        System.out.println("RAM: " + RAM_GB + " GB");
        System.out.println("¿Es gamer?: " + (esGamer ? "Sí" : "No"));
        System.out.println("Precio: €" + getPrecio());
        System.out.println("Stock: " + getStock());
        System.out.println("ID: " + getIdProducto());
    }

    @Override
    public void actualizarPrecio(double porcentaje) {
        double nuevoPrecio = getPrecio() * (1 + porcentaje / 100);
        setPrecio(nuevoPrecio);
    }

    // Implementación de interfaz ConDescuento
    @Override
    public void aplicarDescuento(double porcentaje) {
        double precioConDescuento = getPrecio() * (1 - porcentaje / 100);
        setPrecio(precioConDescuento);
    }

    @Override
    public void mostrarPrecioConDescuento() {
        System.out.println("Precio con descuento aplicado: €" + getPrecio());
    }
}